src/main/webapp/app/components/tabs/curation-data-import-tab/CurationDataImportTab.tsx

```diff
@@ -0,0 +1,553 @@
+}[keyof T];
+
+type OptionalKeys<T> = {
+  [K in keyof T]-?: object extends Pick<T, K> ? K : never;
+}[keyof T];
+
+type DataImportTypeSelectOption = {
+  value: DataImportType;
+  label: string;
+};
+
+enum DataImportType {
+  GENE_GENOMIC_INDICATOR,
+  GENE_SUMMARY,
+  GENE_BACKGROUND,
+  SOMATIC_MUTATION,
+  GERMLINE_MUTATION,
+}
+
+export class DataImportStatus {
+  status: ImportStatus = 'initialized';
+  message: string = '';
+}
+
+const getSelectOption = (dataImportType: DataImportType): DataImportTypeSelectOption => {
+  return {
+    value: dataImportType,
+    label: dataImportTypesLabel[dataImportType],
+  };
+};
+const getSelectOptions = (isGermline: boolean): OptionsOrGroups<any, GroupBase<any>> => {
+  if (isGermline) {
+    return [
+      {
+        label: 'Gene Level',
+        options: [DataImportType.GENE_SUMMARY, DataImportType.GENE_BACKGROUND, DataImportType.GENE_GENOMIC_INDICATOR].map(getSelectOption),
+      },
+      {
+        label: 'Mutation Level',
+        options: [DataImportType.GERMLINE_MUTATION].map(getSelectOption),
+      },
+    ];
+  } else {
+    return [
+      {
+        label: 'Gene Level',
+        options: [DataImportType.GENE_SUMMARY, DataImportType.GENE_BACKGROUND].map(getSelectOption),
+      },
+      {
+        label: 'Mutation Level',
+        options: [DataImportType.SOMATIC_MUTATION].map(getSelectOption),
+      },
+    ];
+  }
+};
+const dataImportTypesLabel: { [key in DataImportType]: string } = {
+  [DataImportType.GENE_GENOMIC_INDICATOR]: 'Genomic Indicator',
+  [DataImportType.GENE_SUMMARY]: 'Summary',
+  [DataImportType.GENE_BACKGROUND]: 'Background',
+  [DataImportType.SOMATIC_MUTATION]: 'Mutation',
+  [DataImportType.GERMLINE_MUTATION]: 'Mutation',
+};
+
+export type DataImportObj = 'hugo_symbol' & GenomicIndicatorDI & GenericDI & SomaticMutationDI & GermlineMutationDI;
+
+const getRequiredColumns = (dataType: DataImportType): RequiredKeys<DataImportObj>[] => {
+  switch (dataType) {
+    case DataImportType.GENE_GENOMIC_INDICATOR:
+      return ['hugo_symbol', 'genomic_indicator'];
+    case DataImportType.GENE_SUMMARY:
+    case DataImportType.GENE_BACKGROUND:
+      return ['hugo_symbol', 'value'];
+    case DataImportType.SOMATIC_MUTATION:
+      return ['hugo_symbol', 'alteration', 'oncogenicity'];
+    case DataImportType.GERMLINE_MUTATION:
+      return ['hugo_symbol', 'alteration', 'pathogenicity'];
+    default:
+      return [];
+  }
+};
+
+const getOptionalColumns = (dataType: DataImportType): OptionalKeys<DataImportObj>[] => {
+  switch (dataType) {
+    case DataImportType.GENE_GENOMIC_INDICATOR:
+      return ['description', 'allele_state'] as OptionalKeys<GenomicIndicatorDI>[];
+    case DataImportType.SOMATIC_MUTATION:
+      return ['name', 'protein_change', 'mutation_effect', 'description'];
+    case DataImportType.GERMLINE_MUTATION:
+      return ['name', 'protein_change', 'description'];
+    default:
+      return [];
+  }
+};
+
+const saveDataToFirebase: {
```

Comment:
@calvinlu3 I don't really like how this method is written. I have some parameters that are only appliable to some of the DataImportType.

Reply:
Also cc @jfkonecn

Reply:
Do you have any suggestions?

Reply:
I'm assuming you want a type error if you add a `DataImportType` so with that in mind you can do something like this.
```ts
type GeneGenomicIndicatorFunction = (
  firebaseGeneService: FirebaseGeneService,
  isGermline: boolean,
  data: DataRow<DataImportObj>
) => Promise<DataRow<DataImportObj>>;

type GeneSummaryFunction = (
  firebaseGeneService: FirebaseGeneService,
  firebaseGeneReviewService: FirebaseGeneReviewService,
  isGermline: boolean,
  data: DataRow<DataImportObj>,
  mutationList: Mutation[],
  vusList: VusObjList
) => Promise<DataRow<DataImportObj>>;

type DefaultFunction = (
  authStore: AuthStore,
  firebaseGeneService: FirebaseGeneService,
  firebaseGeneReviewService: FirebaseGeneReviewService,
  firebaseMetaService: FirebaseMetaService,
  isGermline: boolean,
  data: DataRow<DataImportObj>,
  mutationList: Mutation[],
  vusList: VusObjList
) => Promise<DataRow<DataImportObj>>;

type SaveDataToFirebaseMap = {
  [key in DataImportType]: key extends DataImportType.GENE_GENOMIC_INDICATOR
    ? GeneGenomicIndicatorFunction
    : key extends DataImportType.GENE_SUMMARY
    ? GeneSummaryFunction
    : DefaultFunction;
};
```

Reply:
Thank you! I ended up doing the following based on ur solution
```
type GenomicIndicatorSaveDataFunc = (
  firebaseGeneService: FirebaseGeneService,
  isGermline: boolean,
  data: DataRow<GenomicIndicatorDI>,
) => Promise<DataRow<GenomicIndicatorDI>>;

type GenericGeneDataSaveDataFunc = (
  firebaseGeneService: FirebaseGeneService,
  firebaseGeneReviewService: FirebaseGeneReviewService,
  isGermline: boolean,
  data: DataRow<GenericDI>,
) => Promise<DataRow<GenericDI>>;


type DataImportTypeSaveDataFunc = {
  [DataImportType.GENE_GENOMIC_INDICATOR]: GenomicIndicatorSaveDataFunc;
  [DataImportType.GENE_SUMMARY]: GenericGeneDataSaveDataFunc;
  [DataImportType.GENE_BACKGROUND]: GenericGeneDataSaveDataFunc;
};

const saveDataToFirebase: {
  [key in DataImportType]: key extends keyof DataImportTypeSaveDataFunc ? DataImportTypeSaveDataFunc[key] : null
}

```

Reply:
The things you have to do for type safety 😄